Yosemite/Yosemite/Stores/TaxClassesStore.swift

```diff
@@ -0,0 +1,153 @@
+            }
+        }
+    }
+
+    /// Deletes all of the Stored TaxClasses.
+    ///
+    func resetStoredTaxClasses(onCompletion: () -> Void) {
+        let storage = storageManager.viewStorage
+        storage.deleteAllObjects(ofType: Storage.TaxClass.self)
+        storage.saveIfNeeded()
+        DDLogDebug("Tax Classes deleted")
+
+        onCompletion()
+    }
+
+    /// Synchronizes the Tax Class found in a specified Product.
+    ///
+    func requestMissingTaxClasses(for product: Product, onCompletion: @escaping ([TaxClass]?, Error?) -> Void) {
+        guard let taxClassFromStorage = product.taxClass else {
+            onCompletion(nil, nil)
+            return
+        }
+
+        let storage = storageManager.viewStorage
+        let storageTaxClass = storage.loadTaxClass(slug: taxClassFromStorage)
+
+        if storageTaxClass == nil {
+            let remote = TaxClassesRemote(network: network)
+            remote.loadAllTaxClasses(for: Int64(product.siteID)) { [weak self] (taxClasses, error) in
+                guard let taxClasses = taxClasses else {
+                    onCompletion(nil, error)
+                    return
+                }
+
+                self?.upsertStoredTaxClassesInBackground(readOnlyTaxClasses: taxClasses) {
+                    onCompletion(taxClasses, nil)
+                }
+            }
+        }
+        else {
+            guard let taxClass = storageTaxClass?.toReadOnly() else {
+                onCompletion(nil, nil)
+                return
+            }
+            onCompletion([taxClass], nil)
+            return
+        }
+    }
+}
+
+
+// MARK: - Storage: TaxClass
+//
+private extension TaxClassesStore {
+
+    /// Updates (OR Inserts) the specified ReadOnly TaxClass Entities *in a background thread*. onCompletion will be called
+    /// on the main thread!
+    ///
+    func upsertStoredTaxClassesInBackground(readOnlyTaxClasses: [Networking.TaxClass], onCompletion: @escaping () -> Void) {
+        let derivedStorage = sharedDerivedStorage
+        derivedStorage.perform {
+            self.upsertStoredTaxClasses(readOnlyTaxClasses: readOnlyTaxClasses, in: derivedStorage)
+        }
+
+        storageManager.saveDerivedType(derivedStorage: derivedStorage) {
+            DispatchQueue.main.async(execute: onCompletion)
+        }
+    }
+
+    /// Updates (OR Inserts) the specified ReadOnly TaxClass Entities into the Storage Layer.
+    ///
+    /// - Parameters:
+    ///     - readOnlyTaxClasses: Remote TaxClass to be persisted.
+    ///     - storage: Where we should save all the things!
+    ///
+    func upsertStoredTaxClasses(readOnlyTaxClasses: [Networking.TaxClass], in storage: StorageType) {
+        for readOnlyTaxClass in readOnlyTaxClasses {
+            let storageTaxClass = storage.loadTaxClass(slug: readOnlyTaxClass.slug) ?? storage.insertNewObject(ofType: Storage.TaxClass.self)
+
+            storageTaxClass.update(with: readOnlyTaxClass)
+        }
```

Comment:
we might also want to delete any models in storage but not in the `readOnlyTaxClasses` array in case some of them are deleted on the server. since there's no paging in the API request, we could probably just delete all tax classes before inserting from the API models by `storage.deleteAllObjects(ofType: Storage.TaxClass.self)`, or deleting any storage model whose slug isn't in the API models